### ISSUE TITLE
Always succeed empty fragment checks

### ISSUE THREAD

**Issue by MichaIng** [lycheeverse/lychee#1599]
Currently we see errors like
```
file:///home/runner/work/repo/path/to/stats.html# | Failed: Cannot find fragment
```
while the empty fragment `#` is well known to scroll to the top of the page, just like `#top`, while the latter can be overwritten, if there is an HTML element with `id="top"`. I suggest to succeed the fragment check right away if any of those two special fragments is found.

Sadly I could not find a good source in RFCs or similar. Maybe it is just browsers implementing this outside of specs. The empty fragment is very widely used, the `#top` however I found just now, while searching for a related source. It works however well in all browsers I checked: https://stackoverflow.com/a/77560206

**Comment by thiru-appitap** [comment]
Please apply this patch and confirm if this resolves the issue
[top_fragment.patch](https://github.com/user-attachments/files/18247304/top_fragment.patch)

FYI: the patch also updates the fixtures/fragments/file.html with the empty fragment anchor and a #top anchor (included for testing)

**Comment by MichaIng** [comment]
@thiru-appitap 
Somehow it does not work here. Does it succeed the added tests in your case? `cargo test` currently runs into OOM errors here 😅.
```console
root@VM-Bookworm:~/DietPi-Docs# grep '"top"' /root/lychee/lychee-lib/src/utils/fragment_checker.rs
                let contains_fragment = fragment.is_empty() && file_frags.contains("top")
root@VM-Bookworm:~/DietPi-Docs# /root/lychee/target/debug/lychee --include-fragments -vvv test.md
     [200] file:///root/DietPi-Docs/build/docs/index.html#getting-started
   [ERROR] file:///root/DietPi-Docs/build/docs/index.html#nonexistent | Cannot find fragment
   [ERROR] file:///root/DietPi-Docs/build/docs/index.html# | Cannot find fragment
   [ERROR] file:///root/DietPi-Docs/build/docs/index.html#top | Cannot find fragment

Issues found in 1 input. Find details below.

[test.md]:
   [ERROR] file:///root/DietPi-Docs/build/docs/index.html#nonexistent | Cannot find fragment
   [ERROR] file:///root/DietPi-Docs/build/docs/index.html#top | Cannot find fragment
   [ERROR] file:///root/DietPi-Docs/build/docs/index.html# | Cannot find fragment
```

**Comment by mre** [comment]
Can you open a pull request with your last attempt and the changes from @thiru-appitap? Then I can take a look and perhaps spot the bug.

**Comment by MichaIng** [comment]
PR up: #1609
Tests included.

**Comment by mre** [comment]
Thanks for the PR. The race-condition you describe is certainly odd.

**Comment by MichaIng** [comment]
Just to be sure: any cache includes the fragment, right?

**Comment by mre** [comment]
The fragment cache is here: https://github.com/lycheeverse/lychee/blob/master/lychee-lib/src/utils/fragment_checker.rs
The cache is stored in a `HashMap` with the URL as the key and a `HashSet` of fragments as the value.

The normal request caching logic is here: https://github.com/lycheeverse/lychee/blob/a11d51549b3d1b9fd1be2247bee1588fc01a751a/lychee-bin/src/commands/check.rs#L270-L310
The full URI (including the fragment) is cached.

**Comment by thiru-appitap** [comment]
@MichaIng, @mre I was away and so have missed this conversation so far - my comments below:

1. yes - the simplified logic check (EDIT: 2) makes more sense.
2. The empty fragment (#) check is failing because:
    * in the cache-hit scenario, empty fragment check is missing (fragment.is_empty()
(extractor never includes the empty string for "#" into the fragments list and so an explicit empty fragment check is required)

Made this below code change and verified it is working (file1.html & file2.md changes are noted below)
```
       let is_emtpty_or_top_fragment = fragment.is_empty() || fragment.eq_ignore_ascii_case("top");
        match self.cache.lock().await.entry(url_without_frag) {
            Entry::Vacant(entry) => {
                let content = fs::read_to_string(path).await?;
                let file_frags = extractor(&content);
                let contains_fragment = is_emtpty_or_top_fragment 
                    || file_frags.contains(fragment)
                    || file_frags.contains(&fragment_decoded as &str);
                log::debug!("contains_fragment ({fragment}): {contains_fragment}");
                entry.insert(file_frags);
                Ok(contains_fragment)
            }
            Entry::Occupied(entry) => Ok(is_emtpty_or_top_fragment
                || entry.get().contains(fragment)
                || entry.get().contains(&fragment_decoded as &str)),
        }
```
3. Testing:
    a). made below changes into file.html directly and tested with it to confirm the fix is working
```
  <body>
    <p id="top">Top of the page</p>
    <section id="in-the-beginning" style="height: 100vh;">
      <p id="Upper-ÄÖö">
        <div id="tangent%3A-kustomize"></div>
        To start
        <a href="file1.md#fragment-1">
          let's run away.
        </a>
      </p>
    </section>
    <section>
      <p id="a-word">Word</p>
      <a href="#in-the-beginning">back we go</a><br>
      <a href="#in-THE-begiNNing">back we go upper does not work</a><br>
      <a href="#tangent%3A-kustomize">id with percent encoding</a><br>
      <a href="#Upper-ÄÖö">back to Upper-ÄÖö</a><br>
      <a href="#Upper-%C3%84%C3%96%C3%B6">back to öüä encoded</a><br>
      <a href="#in-the-end">doesn't exist</a><br>
      <a href="#">To the top through empty fragment</a><br>
      <a href="#top">Back to the top</a>
      <a href="#embeddedtopabc">To top (embedded in between) - does not work</a>
      <a href="#endingtop">To top (ending) - does not work</a>
    </section>
  </body>
```
    b). added an empty heading (as below) to file1.md and confirmed working of file1.md# & file1.md#top (browser honours both of these)
```
#  

```

BTW, the cargo test had succeeded in my case and i didn't encounter the race condition - have run few tests and still couldn't hit the issue and will keep trying.

**Comment by mre** [comment]
If you don't mind, could you open a pull request with your changes similar to how @MichaIng did? It would make review and testing much easier.

**Comment by MichaIng** [comment]
@thiru-appitap
Great find with the missing empty/top fragment handling, when a cache entry is present. This explains the race condition/inconsistent results in the same turn, since checks run in parallel, and sometimes the cache key might have been created already, sometimes not.

Why did you add a `top` ID and an empty heading to the test files? Those fragments are supposed to work without any of the IDs. Hence in my tests I added checks for those fragments, but did explicitly not add related headings or IDs, otherwise false negatives are possible.

**Comment by thiru-appitap** [comment]
> @thiru-appitap Great find with the missing empty/top fragment handling, when a cache entry is present. This explains the race condition/inconsistent results in the same turn, since checks run in parallel, and sometimes the cache key might have been created already, sometimes not.
> 
> Why did you add a `top` ID and an empty heading to the test files? Those fragments are supposed to work without any of the IDs. Hence in my tests I added checks for those fragments, but did explicitly not add related headings or IDs, otherwise false negatives are possible.

@MichaIng 
You are correct - I should've removed those entries (`top` ID and empty heading) from the test files (and doing so in my fork :-)) - thanks for pointing it out!

**Comment by mre** [comment]
Thanks for the great teamwork! This is now fixed in `master` and planned to be released soon.
